src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ISetOfTConverter.cs

```diff
@@ -11,7 +11,7 @@ internal sealed class ISetOfTConverter<TCollection, TElement>
     {
         protected override void Add(in TElement value, ref ReadStack state)
         {
-            ((TCollection)state.Current.ReturnValue!).Add(value);
+            ((ISet<TElement>)state.Current.ReturnValue!).Add(value);
```

Comment:
Given the below benchmark results (run with [BenchmarkDotNet](https://benchmarkdotnet.org/articles/overview.html)), there is potentially a perf cost with this change.

To take this PR, we'd want to see before/after results for the [JSON collection deserialization benchmarks](https://github.com/dotnet/performance/blob/a3895c72f9dda4904db78871b121a1f062611894/src/benchmarks/micro/libraries/System.Text.Json/Serializer/ReadJson.cs#L30). This [document](https://github.com/dotnet/performance/blob/a3895c72f9dda4904db78871b121a1f062611894/docs/benchmarking-workflow-dotnet-runtime.md) describes how to run them. 

To mitigate any regressions, I'd consider the suggestion in https://github.com/dotnet/runtime/pull/41223#discussion_r475870470 again to limit the changes to just value-type collections.

It may be worth it to see if using [`Unsafe.Unbox`](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.unbox?view=net-5.0) provides better perf for value-type scenarios as suggested in https://github.com/dotnet/runtime/issues/1998#issuecomment-678762563. As the API is not available on .NET Framework or .NET Standard, the way to use it would be:

```cs
#if NETCOREAPP
// Use Unsafe.Unbox API
#else
// Do something else.
#endif
```

I'd also measure the perf for the approach in https://github.com/dotnet/runtime/pull/41223#discussion_r475870470.

---

|     Method | Count |        Mean |      Error |     StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |------ |------------:|-----------:|-----------:|------:|--------:|------:|------:|------:|----------:|
| ActualType |    10 |    49.64 ns |   0.169 ns |   0.141 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|  Interface |    10 |    83.10 ns |   1.684 ns |   1.802 ns |  1.68 |    0.04 |     - |     - |     - |         - |
|            |       |             |            |            |       |         |       |       |       |           |
| ActualType |    50 |   263.25 ns |   1.526 ns |   1.353 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|  Interface |    50 |   384.83 ns |   3.110 ns |   2.757 ns |  1.46 |    0.01 |     - |     - |     - |         - |
|            |       |             |            |            |       |         |       |       |       |           |
| ActualType |   500 | 2,481.30 ns |  15.786 ns |  13.994 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|  Interface |   500 | 3,784.60 ns |  31.093 ns |  29.084 ns |  1.52 |    0.01 |     - |     - |     - |         - |
|            |       |             |            |            |       |         |       |       |       |           |
| ActualType |  1000 | 6,281.58 ns |  53.583 ns |  50.121 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|  Interface |  1000 | 7,679.08 ns | 139.002 ns | 130.023 ns |  1.22 |    0.02 |     - |     - |     - |         - |

```cs
[MemoryDiagnoser]
public class PopulateISetOfT
{
    private object _collection;

    [Params(10, 50, 500, 1000)]
    public int Count;

    [GlobalSetup]
    public void Setup()
    {
        _collection = new HashSet<int>();
    }

    [Benchmark(Baseline = true)]
    public void ActualType()
    {
        for (int i = 0; i < Count; i++)
        {
            ((HashSet<int>)_collection).Add(i);
        }
    }

    [Benchmark]
    public void Interface()
    {
        for (int i = 0; i < Count; i++)
        {
            ((ISet<int>)_collection).Add(i);
        }
    }
}
```

Reply:
>As the API is not available on .NET Framework or .NET Standard

@layomia Regardless of what I thought earlier I think `Unsafe.Unbox` is available on all target frameworks. The package is already a dependency of S.T.J. on `netstandard2.0` and `net461`. On `netcoreapp3.0` it's inbox. It works on all platforms as expected (I used UWP to check `netstandard2.0`)

Reply:
Anyway it's not necessary here. This does the job;
```c#
            var collection = (TCollection)state.Current.ReturnValue!;
            collection.Add(value);
            if (IsValueType)
            {
                state.Current.ReturnValue = collection;
            };
```
I assume it performs slightly better but using `Unsafe.Unbox<T>()`  would require to duplicate all affected converters because `Unbox` has a `where T: struct` restriction.